test/testutils/client.go

```diff
@@ -0,0 +1,40 @@
+package testutils
+
+import (
+	"context"
+	"testing"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DeleteObject deletes obj from c.
+// Succeeds on success or if obj already did not exist in c.
+// Fails the test if calling Delete() returns an error other than NotFound.
+//
+// Does not ensure the object is actually deleted, just sends a delete request via the passed Client.
+// See #1596 for why this is non-trivial to guarantee.
+//
+// Tests should not rely on objects being deleted from the API Server which were used in other tests; they should
+// instead use unique object names and namespaces to eliminate cross-talk.
+func DeleteObject(t *testing.T, c client.Client, original client.Object) func() {
+	// We don't want this cleanup method to rely on any context passed by the caller. For example, the caller may have
+	// canceled their context as part of their test.
+	ctx := context.Background()
+
+	objCopy := original.DeepCopyObject()
+	obj, ok := objCopy.(client.Object)
+	if !ok {
+		t.Fatalf("got DeepCopyObject(%T) = %T, which is not a client.Object", original, objCopy)
```

Comment:
> It _shouldn't_. The problem is that:
> 
> 1. Go doesn't have generics, and
> 2. The client.Object's `DeepCopyObject()` method returns a `runtime.Object`.
> 
> It's possible for someone to define their `client.Object`-satisfying type that returns a `runtime.Object` of a different type when told to `DeepCopyObject()` which is a `runtime.Object` but not a `client.Object`.

Would I do this by manually overriding the generated deepcopy functions?  Or is there another way of achieving that same behavior?

Reply:
> The deepcopy here is because the DeleteObject method only accepts a _reference_ to the object. It is possible that the caller then uses a method such as `obj.SetName()` which would result in a different object being cleaned up.
> 
> I originally just copied out the GVK and Namespace/Name, but the problem is that many tests don't set GVK on struct types, so without a schema it isn't possible to get the GVK from the struct.

So people are changing the name of an object, passing it to `c.Delete`, then expecting the _original_ object to be cleaned up instead of the _new_ object with the different name??

Reply:
I don't see any cases of that happening. This is a safeguard _in case_ someone does something like that.